proposals/0000-arkworks-integration.md

```diff
@@ -0,0 +1,60 @@
+---
+Title: Arkworks Integration
+Number: 0
+Status: Proposed
+Authors:
+  - Achim Schneider<achim@parity.io>
+Created: 2023-13-03
+Category: Runtime Environment
+Requires:
+Replaces:
+--- 
+
+## Summary
+ Proposal to merge https://github.com/paritytech/substrate/pull/13031 into Substrate. [Arkworks](http://arkworks.rs/) integration into Substrate. We provide Elliptic curves (bls12_381, bls12_377, ed_on_bls12_381, ed_on_bls12_77, bw6_761), were we replace the compute intense operations by host function calls in Substrate. We add those host function calls to `sp_io::elliptic_curves` and refer to the `arkworks` crate. 
+The curves live in https://github.com/paritytech/ark-substrate and call into the host functions. To avoid point preparation of the elliptic curves in the runtime, we also partially fork the models `bls12` and `bw6` in [ark-substrate](https://github.com/paritytech/ark-substrate).
+
+## Motivation
+Usually cryptographic operations on elliptic curves are slow in WebAssembly. Replacing those operations by host functoin calls into binary code allows us to drastically improve the performace of the cryptographic tools which rely on those operations.
+
+## Detailed Solution design
+We implement host function calls into the underlying arithmetic operations on elliptic curves for `BLS12_377`, `BLS12_381`, `BW6_761`, `ED_ON_BLS12_377` and `ED_ON_BLS12_381` by host function calls in Substrate.
+
+We introduce new host functions calls to Substrate which are grouped under `elliptic_curves` in `/primitives/io/src/lib.rs`. Those host function calls receive serialized curve points of elliptic curves, deserialize them, call into native arkworks code to perform computations. Finally they serialize the result and return them as `Vec<u8>`.
+
+While the host-functions themself are implemented under `/primitives/io/src/lib.rs`, they call into their actual implementations which can be found in the `/primitives/io/arkworks` sub-crate.
+
+Those host functions calls are used by the [ark-substrate]() library. When using elliptic curves from ark-substrate, one passes the required host-function by depdendency injection. This allows for a clean separation of concerns and to keep the amount of code contributed to the Substrate core repo as minimal as possible.
+
+For testing purposes, we are able to re-use the extensive arkworks testing macros, those tests are performed under `/primitives/arkworks/tests` folder.
+
+Example implementations and benchmarks for the code can be found under: https://github.com/achimcc/substrate-arkworks-examples.
+
+### Benchmarking results
+The following comparison table of benchmarking results should justify the introduction of the new host functions to Substrate:
+
+ ed_on_bls12_381_msm_te, 10 arguments        |    7813.27       |    3207.47      |    35.21        |     12470       |      560.82     |
```

Comment:
What does this table show? What are the headings?

Reply:
Sorry, was a copy & paste mistake, is fixed!